### ISSUE TITLE
Function to perform non-strict conversion of bytes to [target encoding]

### ISSUE THREAD

**Issue by heyciao** [cockroachdb/cockroach#66916]
General infos:
- CockroachDB (CRDB) version 21.1.2
- DBeaver version 21.1.0.202106012023
- I'm an absolute beginner with CRDB & in general with character encoding => not sure if such a conversion can already be done in CRDB.

This was initially discussed [here](https://forum.cockroachlabs.com/t/non-strict-conversion-of-bytes-to-utf8-string/4650).

**Current situation**
When trying to convert a sequence of bytes to UTF-8 by using the function "convert_from()" the CRDB returns an error message when any byte(s) is(are) invalid for the target encoding and/or when the length of the sequence of bytes is invalid.

**Wished improvement**
A "best-effort" function that never fails, which performs the conversion to the target encoding for valid sequences of bytes but that just prints some placeholder for invalid byte(s) of sequences of bytes.
Maybe better than creating a brand new function: just an additional parameter integrated in the existing function(s) that enables/disables the "best-effort"/"lax" conversion.

**Reason**
The data that I'm storing in my "BYTES"-column can be anything - pure ascii text, original UTF-8 text, text converted to UTF-8 from other encodings, pure binary data, mixed binary data and text, so basically anything - and it can be as well suddenly truncated at any length.
My programs use only the data's binary values (so no problems here), but to do manual healthchecks/investigations on the data, being able to look at all its forms (ascii + hexed + interpreted which is UTF-8 in my case) is a huge help for me.

I guess that this would be useful as well to other users that e.g. might have data which is only 99.99...% correct.
Errors during data download/read/transformation/etc... can happen at any level and not being able to perform at all the conversion just because a single byte within thousands of bytes/lines is wrong is a bit extreme => it would be nice to be able to perform a "lax" conversion.

**Details - example 1**
I'm going to use 2 sequences of bytes, each byte of each sequence written here as hex:
1. the correct sequence "68c3bc68" which is supposed to print the chars "hüh" (the hex 68 for the "h" and the hex "c3+bc" for the UTF-8 german char "ü" like "München", "Müller", "Brücke", etc... - you should be able to see it [here](https://www.utf8-chartable.de/) on the 4th last line)
2. the wrong sequence "68c3bcc368c3" which is similar to the previous example but which contains the byte "c3" at 2 different locations not followed by an additional correct byte.

Converting (in DBeaver) the **correct** sequence (nbr 1) to UTF-8:
- MariaDB (first convert the hex to bytes using "unhex" and then to UTF-8 using "convert"):
`select CONVERT(unhex('68c3bc') USING utf8)`
=> result is "hü" => perfect.
- Clickhouse (don't know, it just works):
`select unhex('68c3bc')`
=> result is "hü" => perfect.
- CRDB (first convert the hex to bytes using "decode" and then to UTF-8 using "convert_from")::
`select convert_from( decode('68c3bc', 'hex'), 'utf8')`
=> result is "hü" => perfect.

Converting (in DBeaver) the **wrong** sequence (nbr 2) to UTF-8:
- MariaDB:
`select CONVERT(unhex('68c3bcc368c3') USING utf8)`
=> result is "hü?h?" => acceptable. The conversion worked for the valid ascii and the single utf-8 byte-sequence, but the invalid bytes were converted into ascii question marks ("?") which makes it impossible to differenciate against bytes which had their values set for real to be a question mark.
- Clickhouse:
`select unhex('68c3bcc368c3')`
=> result is "hü�h�" => quite good. The invalid bytes were converted to "�" which is definitely not a normal question mark char => it distinguishes itself clearly from normal ascii chars => better than MariaDB.
- CRDB:
`select convert_from( decode('68c3bcc368c3', 'hex'), 'utf8')`
=> result is the error message "SQL Error [22021]: ERROR: convert_from(): invalid byte sequence for encoding "UTF8"" => technically correct, but for me useless :P

**Details - example 2**
To recap, the original sequence of correct bytes was "68c3bc" which make up the UTF-8 chars "hü".
I'm going now to add to it a single additional invalid byte to it, being "f" => the sequence of bytes becomes therefore "68c3bcf" => this basically "moves" all existing bytes by one position, changing all of their interpreted values.
- MariaDB:
`select CONVERT(unhex('68c3bcf') USING utf8)`
=> result is "?;?" => not visible here but the first char is a funny thing that shows the hex 0006 => all good for me with again the weird situation of the ascii-question-marks.
- Clickhouse:
`select unhex('68c3bcf')`
=> result is "�;�" => basically the same thing like MariaDB - apparently not visible here but the first char is again the funny thing that shows the hex 0006, then the funny question mark + semicolon + funny question mark ("�;�") => better than MariaDB because it does not use the ascii-question-marks.
- CRDB:
`select convert_from( decode('68c3bcf', 'hex'), 'utf8')`
=> result is the error message "SQL Error [XXUUU]: ERROR: decode(): encoding/hex: odd length hex string" => I know that the sequence of bytes does not comply to the standard => not useful.

I'm personally not sure why MariaDB&Clickhouse decide to show sometimes the replacement char (plain ascii question mark "'?" for MariaDB and "�" in the case of Clickhouse) or the tiny-char showing the hex (not sure if it's really the hex) of the bytes.
Personally I would just always display the hex (**as a single char** - if I remember correctly CRDB can partially show intermediate UTF-8 sequences but the other chars are escaped and that totally screws up the whole alignment between rows), but maybe the people at MariaDB/Clickhouse are doing that for a specific reason?

Thank you
Stefano

Jira issue: CRDB-8286

**Comment by rafiss** [comment]
Thanks for the detailed request @heyciao! 

I have a question -- you mentioned that the data you are dealing with could be _anything_ -- then why do you prefer to decode it into a UTF8 string? Would it be more useful to simply view the hex encoding or base64 encoding? For example, if the data really was just binary data, then wouldn't it be meaningless to look at the UTF8 encoding?

**Comment by heyciao** [comment]
Hi @rafiss 
Well, it depends... .

Generally speaking I do associate the data that I insert into the DB to an identifier that is supposed to tell me in general what's the type of data that I'm dealing with (e.g. executable code, text, video, some kind of mix, etc...) but that identifier can be sometimes wrong (e.g. in Internet files can be associated to a wrong mime-type, or my autorecognition might return a wrong result, etc...).
Therefore when having e.g. 10 rows that are supposed to contain text but if 1 would be associated to a wrong type then I would run into the error message when trying to show them as UTF-8. Or maybe that "bad" row might be indeed text, but encoded with e.g. the japanese Shift-JIS which would generate as well the same error message when trying to show it as UTF-8.
Basically, if the data is not perfect I get always just get the error message.

Concerning binary & mixed data, displaying it as UTF-8 just makes things easier than using only hex/base64 even if the result is then 99.9999% garbage (but at least similar rows return similar garbage). Some sequences of bytes might contain as well text (e.g. metadata stored in some audio/video files can be complex, like the names of the songs together with the cover picture of the song) so here it's just nice to be able to directly identify from such metadata the text even if it's mixed with binary data.
